k8s/src/main/scala/io/buoyant/k8s/resources.scala

```diff
@@ -0,0 +1,308 @@
+
+  /**
+   * version within owner domain, i.e. "v1"
+   */
+  def ownerVersion: String
+
+  def version: String = ThirdPartyVersion.version(owner, ownerVersion)
+  def basePath: String = s"/${ThirdPartyVersion.group}/$version"
+  val group = ThirdPartyVersion.group
+  override def namespace(ns: String) = new NsThirdPartyVersion[O](client, owner, ownerVersion, ns)
+}
+
+object ThirdPartyVersion {
+  def version(owner: String, ownerVersion: String) = s"$owner/$ownerVersion"
+  val group = "apis"
+}
+
+/**
+ * A namespaced third-party version, i.e. "/apis/buoyant.io/v1/namespaces/ns/".
+ */
+class NsThirdPartyVersion[O <: Object](client: Client, owner: String, ownerVersion: String, ns: String)
+  extends NsVersion[O](client, ThirdPartyVersion.group, ThirdPartyVersion.version(owner, ownerVersion), ns)
+
+/**
+ * A trait that encapsulates the ability to Watch a k8s resource.
+ */
+private[k8s] trait Watchable[O <: Object, W <: Watch[O]] extends Resource {
+  implicit protected val watchManifest: Manifest[W]
+
+  protected def backoffs: Stream[Duration]
+  protected def path: String
+  protected def stats: StatsReceiver
+
+  /**
+   * implementing classes should define this method to retrieve the current state
+   * of the resource to pass as Modified results to any watchers.
+   *
+   * @return a Future containing a sequence of Watches and an optional String representing the current resourceVersion
+   */
+  protected def restartWatches(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None
+  ): Future[(Seq[W], Option[String])]
+
+  protected def infiniteRetryFilter = new RetryFilter[http.Request, http.Response](
+    RetryPolicy.backoff(backoffs) {
+      // We will assume 5xx are retryable, everything else is not for now
+      case (_, Return(rep)) => rep.status.code >= 500 && rep.status.code < 600
+      case (_, Throw(NonFatal(ex))) =>
+        log.error(s"retrying k8s endpoints request on error $ex")
+        true
+    },
+    HighResTimer.Default,
+    stats,
+    RetryBudget.Infinite
+  )
+
+  /**
+   * Watch this resource for changes, using a chunked HTTP request.
+   *
+   * TODO: investigate k8s websockets support as an alternative HTTP approach
+   */
+  def watch(
+    labelSelector: Option[String] = None,
+    fieldSelector: Option[String] = None,
+    resourceVersion: Option[String] = None
+  ): (AsyncStream[W], Closable) = {
+    val close = new AtomicReference[Closable](Closable.nop)
+
+    // Internal method used to recursively retry watches as needed on failures.
+    def _watch(resourceVersion: Option[String] = None): AsyncStream[W] = {
+      val req = Api.mkreq(http.Method.Get, path, None,
+        "watch" -> Some("true"),
+        "labelSelector" -> labelSelector,
+        "fieldSelector" -> fieldSelector,
+        "resourceVersion" -> resourceVersion)
+
+      def currentVersion(initial: Option[String], watch: AsyncStream[W]): Future[Option[String]] = {
+        watch.foldLeft(initial) {
```

Comment:
this just takes the resourceVersion of the last watch?  maybe define a helper method `def last[A](as: AsyncStream[A]): Future[Option[A]]`